src/job_retry.py

```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env python3
+#
+# SPDX-License-Identifier: LGPL-2.1-or-later
+#
+# Copyright (C) 2025 Collabora Limited
+# Author: Jeny Sadadia <jeny.sadadia@collabora.com>
+
+import sys
+import time
+import kernelci.config
+from kernelci.legacy.cli import Args, Command, parse_opts
+
+from base import Service
+
+
+class JobRetry(Service):
+
+    subscription_filters = {
+        "state": "done",
+        "result": ("fail", "incomplete"),
+        "kind": "job"
+        # ToDo: Retry for build jobs
+        # "kind": ("kbuild", "job")
+    }
+
+    def __init__(self, configs, args):
+        super().__init__(configs, args, 'job_retry')
+
+    def _setup(self, args):
+        return self._api_helper.subscribe_filters(self.subscription_filters)
+
+    def _stop(self, sub_id):
+        if sub_id:
+            self._api_helper.unsubscribe_filters(sub_id)
+        sys.stdout.flush()
+
+    def _find_parent_kind(self, node, api_helper, kind):
+        parent_id = node.get('parent')
+        if not parent_id:
+            return None
+        parent_node = api_helper.api.node.get(parent_id)
+        if not parent_node:
+            return None
+        if parent_node.get('kind') == kind:
+            return parent_node
+        return self._find_parent_kind(parent_node, api_helper, kind)
+
+    def _run(self, sub_id):
+        self.log.info("Job retry: Listening for events... ")
+        self.log.info("Press Ctrl-C to stop.")
+        subscribe_retries = 0
+        while True:
+            try:
+                node, _, _ = self._api_helper.receive_event_node(sub_id)
+                self.log.debug(f"Event received: {node['id']}")
+            except Exception as e:
+                self.log.error(f"Error receiving event: {e}, re-subscribing in 10 seconds")
+                time.sleep(10)
+                sub_id = self._api_helper.subscribe_filters(self.subscription_filters)
+                subscribe_retries += 1
+                if subscribe_retries > 3:
+                    self.log.error("Failed to re-subscribe to node events")
+                    return False
+                continue
+            subscribe_retries = 0
+
+            # Check retry count before submitting a retry
+            retry_counter = node.get("retry_counter", 0)
+            self.log.debug(f"{node['id']}: Node current retry_counter: {retry_counter}")
+            if retry_counter >= 3:
+                self.log.info(f"{node['id']} Job has already retried 3 times. \
+Not submitting a retry.")
+                continue
+
+            parent_kind = None
+            if node.get('kind') == 'job':
+                parent_kind = 'kbuild'
+            # ToDo: retry build jobs
+            # if node.get("kind") == "kbuild":
+            #     parent_kind = "checkout"
+            if parent_kind:
+                event_data = self._find_parent_kind(node, self._api_helper, parent_kind)
+                event_data["jobfilter"] = [node["name"]]
```

Comment:
```
[ERROR] Traceback (most recent call last):
  File "/home/kernelci/pipeline/src/base.py", line 70, in run
    status = self._run(context)
             ^^^^^^^^^^^^^^^^^^
  File "/home/kernelci/pipeline/./src/job_retry.py", line 80, in _run
    event_data["jobfilter"] = [node["name"]]
    ~~~~~~~~~~^^^^^^^^^^^^^
TypeError: 'NoneType' object does not support item assignment
```
because `if parent_kind:` is not checking all parameter used inside the if function,
in the event of the the retry counter get to 0 and information get mangled the script crash if cannot find such informations

Reply:
Fixed it. Thanks.